[JENKINS-28335] - Git username and password binding (GSoC 2021)

arpoch opened the pull request:
## [JENKINS-28335](https://issues.jenkins.io/browse/JENKINS-28335) - Add Git Credentials binding for Username and Password

The gitUsernamePassword implementation uses the Jenkins username and password values retrieved through Credential API, to access a remote repository over HTTP protocol. The binding uses the GIT_ASKPASS environment variable to provide credentials requested by a git operation in a pipeline job/freestyle project. Executable script(sh,bat,powershell) is attached/binded to GIT_ASKPASS variable, which is invoked when asked for user’s credentials by HTTP server.

Please refer to the Project page for more details-
https://www.jenkins.io/projects/gsoc/2021/projects/git-credentials-binding/

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/jenkinsci/git-plugin/blob/master/CONTRIBUTING.adoc) doc
- [x] I have referenced the Jira issue related to my changes in one or more commit messages
- [x] I have added tests that verify my changes
- [x] Unit tests pass locally with my changes
- [x] I have added documentation as necessary
- [x] No Javadoc warnings were introduced with my changes
- [x] No spotbugs warnings were introduced with my changes
- [x] Documentation in README has been updated as necessary
- [x] Online help has been added and reviewed for any new or modified fields
- [x] I have interactively tested my changes
- [x] Any dependent changes have been merged and published in upstream modules (like git-client-plugin)

## Types of changes

- [x] New feature (non-breaking change which adds functionality)

arpoch committed:
Adding Credential Binding Plugin dependency

arpoch committed:
Consuming Git Client plugin incrementals

arpoch committed:
Adding config.jelly file for GitUsernamePasswordBinding

arpoch committed:
Adding inline help for form elements in config.jelly

arpoch committed:
Adding java interface type to support git-credentials binding

arpoch committed:
Adding DisplayName property for UsernamePassword binding

arpoch committed:
GitUsernamePassword binding support for git-credentials

arpoch committed:
Test cases for GitUsernamePasswordBinding

MarkEWaite commented:
My lightweight first pass interactive tests all passed.  Things that I checked:

* Freestyle job with a Git username / password credential binding on Linux is able to `git fetch --all` from a private repository using a shell step.  Same shell step fails if the credentials binding is not used
* Pipeline job with a Git username / password credential binding on Linux is able to `git fetch --all` from a private repository using a shell step.  Same shell step fails if the withCredentials binding is not used

MarkEWaite commented on src/main/java/jenkins/plugins/git/GitUsernamePasswordBinding.java:
```diff
@@ -0,0 +1,182 @@
+    final private String gitToolName;
+    final private Map<String, String> credMap = new LinkedHashMap<>();
+    private String gitToolExe = null;
+    private boolean unixNodeType;
```
I think that this field should be declared `transient` so that it is not stored in the job definition.  I believe the storage in the field is used to pass a value into setRunEnvironmentVariables but should not be considered persistent across jobs.  It is computed based on the launcher used for the current job and the computed value is used immediately to prepare the environment for the credential binding.
```suggestion
    private transient boolean unixNodeType;
```

MarkEWaite commented on src/main/java/jenkins/plugins/git/GitUsernamePasswordBinding.java:
```diff
@@ -0,0 +1,182 @@
+
+    @Override
+    public void setRunEnvironmentVariables(@NonNull FilePath filePath, @NonNull TaskListener listener) throws IOException, InterruptedException {
+        if (unixNodeType && ((CliGitAPIImpl) getGitClientInstance(gitToolExe,null,
```
Doesn't this risk a class cast exception if the getGitClientInstance returns a JGitAPIImpl instance?

arpoch replied:
As mentioned in this [comment](https://github.com/jenkinsci/git-plugin/pull/1104#issuecomment-877800361) the method `getCliGitTool` is the key to prevent this risk.
Also the reason to create getCliGitTool API was to ensure that the binding will only support CLI Git and no other git impls.

MarkEWaite commented on src/main/java/jenkins/plugins/git/GitUsernamePasswordBinding.java:
```diff
@@ -0,0 +1,182 @@
+              //Hard Coded platform dependent newLine
+            if (this.unixNodeType) {
+                gitEcho = workspace.createTempFile("auth", ".sh");
+                // [#!/usr/bin/evn sh] to be used if required, could have some corner cases
```
```suggestion
                // [#!/usr/bin/env sh] to be used if required, could have some corner cases
```

MarkEWaite commented on src/main/resources/jenkins/plugins/git/Messages.properties:
```diff
@@ -25,3 +25,4 @@ GitSCMSource.DisplayName=Git
 GitStep.git=Git
 within.Repository=Within Repository
 additional=Additional
+GitUsernamePasswordBind.DisplayName=Git Username and Password
```
I don't see any mention of `GitUsernamePasswordBind` anywhere else in the code.  Shouldn't this be removed or replaced with a reference to the current class name?
```suggestion
GitUsernamePasswordBind.DisplayName=Git Username and Password
```

MarkEWaite reviewed (commented):
Initial comments and questions

MarkEWaite commented on src/main/resources/jenkins/plugins/git/GitUsernamePasswordBinding/help-credentialsId.html:
```diff
@@ -0,0 +1,31 @@
+<p>
+<strong>Shell example</strong>
+<pre>
+withCredentials([GitUsernamePassword('my-credentials-id')]) {
```
Needs the `credentialsId` argument explicitly named
```suggestion
withCredentials([GitUsernamePassword(credentialsId: 'my-credentials-id')]) {
```

MarkEWaite commented on src/main/resources/jenkins/plugins/git/GitUsernamePasswordBinding/help-credentialsId.html:
```diff
@@ -0,0 +1,31 @@
+<p>
+<strong>Batch example</strong>
+<pre>
+withCredentials([GitUsernamePassword('my-credentials-id')]) {
```
Needs the `credentialsId` argument explicitly named
```suggestion
withCredentials([GitUsernamePassword(credentialsId: 'my-credentials-id')]) {
```

MarkEWaite commented on src/main/resources/jenkins/plugins/git/GitUsernamePasswordBinding/help-credentialsId.html:
```diff
@@ -0,0 +1,31 @@
+<p>
+<strong>Powershell example</strong>
+<pre>
+withCredentials([GitUsernamePassword('my-credentials-id')]) {
```
Needs the `credentialsId` argument explicitly named
```suggestion
withCredentials([GitUsernamePassword(credentialsId: 'my-credentials-id')]) {
```

MarkEWaite committed:
Make docs match working syntax

MarkEWaite committed:
Spelling fix

arpoch commented:
> Enabling JGit and dragging it to be first in the list of implementations will cause the build log to report that JGit is not supported with the credential binding, even if the git tool name is 'Default' or 'git-2.32.0' (a tool that I defined). We should allow JGit to be selected as the first git implementation as is done on ci.jenkins.io.

The issues is due to this [line](https://github.com/jenkinsci/git-plugin/blob/b4fdcf9a4829ef72a27b37af451a187befabd221/src/main/java/jenkins/plugins/git/GitCredentialBindings.java#L38),  which checks if the installation is `GitTool` type specific and not `JGitTool` or `JGitApache` type specific, the assumption made here of selecting a tool at index 0 is what lead to this situation. 
I have made some fixes and things are back on track.

Since your binding will only work for Cli Git specific git tool so its better to convert the method [getGitTool](https://github.com/jenkinsci/git-plugin/blob/b4fdcf9a4829ef72a27b37af451a187befabd221/src/main/java/jenkins/plugins/git/GitCredentialBindings.java#L28) to `getCliGitTool` returning a git tool which is both cli git type specific and name specific provided by the user. So returning a generic git tool as done by `resloveGitTool` won't sever our entire purpose.

arpoch committed:
Changing property name for GitUsernamePasswordBinding

arpoch committed:
Resolving gitTool return error due to assumed index for ToolInstallation array

The method method signature and implementation have been changed to return Cli Git specific git tool. The methods name have been changed to `getCliGitTool`.

Instead of returning a generic git tool as done by resolveGitTool as it won't sever our entire purpose the `getCliGitTool` method is returning a git tool which is both cli git type specific and name specific provided by the user.

This method ensures that the gitTool available to the class implementing this interface is of type cli git tool.

arpoch committed:
Check for node os type if cli git tool is found

arpoch committed:
Merge branch 'gitUsernamePassword' of https://github.com/arpoch/git-plugin into gitUsernamePassword

MarkEWaite commented:
@arpoch could you merge the master branch into this branch so that the incremental builds generated by ci.jenkins.io will be stored in the Jenkins artifact repository?  That will allow me to insert the incremental version number into my plugins.txt file and use it to generate my Jenkins configuration from code.

arpoch committed:
Merge branch 'jenkinsci:master' into gitUsernamePassword

arpoch committed:
Adding GitSampleRepoRule rule

arpoch committed:
Adding a test parameter

arpoch committed:
Check GitTool instance by type

arpoch committed:
Removing redundant Git instance

arpoch committed:
Assert environment variables in a Freestyle project for all git tool impl

arpoch committed:
Assert environment variables in a Pipeline job for all git tool impl

arpoch committed:
Test getCliGitTool method using Freestyle project

arpoch committed:
Git tool check using isCliGitTool

arpoch committed:
Formatting code

arpoch committed:
Updating Git Username and Password binding snippet with gitToolName filed

arpoch committed:
Adding form validation for GitTool Name field

arpoch committed:
Merge remote-tracking branch 'origin/gitUsernamePassword' into gitUsernamePassword

arpoch committed:
Merge branch 'jenkinsci:master' into gitUsernamePassword

arpoch committed:
Replacing textBox with listBox for git tool name selection

arpoch committed:
Bumping git-client version to 3.8.0

arpoch committed:
Check gitTool name differently in case not cli git

arpoch committed:
Validation support in case no git tool found of the specified type

arpoch committed:
Using public and secret pairs to bind user's credentials

arpoch committed:
Using public and secret pairs to bind git specific environment variables

arpoch committed:
Use publicValues pair bind git specific environment variables

arpoch committed:
Assert username credential is not displayed in build logs

arpoch committed:
Adding javadocs for GitCredentialBindings methods

arpoch committed:
Assert and default gitTool of type GitTool.class in case gitTool not found

arpoch committed:
Changing cliGitTool field scope to local

MarkEWaite commented:
All my functionality checks pass with the latest code changes (as of b1336813ef9ab9f3f827517cc9cb67743eff738b) .  Well done!

I see the pick list of git tools as well and am able to select one.  Very nice.

Unfortunately, when I see the pick list, it also always shows the message that the default git tool will be used, no matter which git tool I select.  Is that a mistaken check of the field contents or something else?

![screencapture-mark-pc2-markwaite-net-8080-job-Hosting-Providers-job-github-job-github-bin-https-withCredentials-windows-pipeline-syntax-2021-07-16-17_44_35](https://user-images.githubusercontent.com/156685/126018298-ed692fb5-c903-4e02-bfb4-02ca4131730e.png)

MarkEWaite commented on src/main/java/jenkins/plugins/git/GitUsernamePasswordBinding.java:
```diff
@@ -0,0 +1,216 @@
+            return Messages.GitUsernamePasswordBinding_DisplayName();
+        }
+
+        public FormValidation doCheckGitToolName(@QueryParameter String toolName){
+            if(toolName == null){
+                return FormValidation.warning("Default Git Tool(Git/JGit) will be used");
+            }else {
+                return FormValidation.ok();
+            }
+        }
+
```
I believe this code can be removed because there is no need to check for valid content when showing a pick list.  The pick list can only contain valid content.
```suggestion
```

MarkEWaite committed:
Remove form content validator

The pick list content will always be valid.

arpoch committed:
Merge branch 'jenkinsci:master' into gitUsernamePassword

MarkEWaite reviewed (approved):
Needs documentation in the README, but otherwise looks ready to merge and release.

The README documentation can be in a separate pull request if needed.

@rishabhBudhouliya and @justinharringa , any concerns from you that should be addressed before this is merged and released?

arpoch committed:
Add image

Git Username and Password binding in a Pipeline job

arpoch committed:
Add image

Git Username and Password binding in a Freestyle project

arpoch committed:
Git Credential Binding specific to Pipeline

Adding doc for Git Username and Password binding in a Pipeline job.

arpoch committed:
Git Binding in Git plugin

Adding doc for Git Username and Password binding in a Pipeline job and a Freestyle project.

arpoch committed:
Change imports to fully qualified

arpoch committed:
Improving javadocs for getGitClientInstance method

arpoch committed:
Merge branch 'jenkinsci:master' into gitUsernamePassword

MarkEWaite commented on README.adoc:
```diff
@@ -51,6 +51,48 @@ The link:https://youtu.be/B_2FXWI6CWg[30 minute video clip] below introduces Mul
+[#credential-binding]
+=== Git Credentials Binding
+
+The git plugin provides `Git Username and Password` binding that allows authenticated git operations over *HTTP* and *SSH* protocols using command line git in a Pipeline job.
```
SSH support is not yet ready.  Let's not mention it yet.
```suggestion
The git plugin provides `Git Username and Password` binding that allows authenticated git operations over *HTTP* and *HTTPS* protocols using command line git in a Pipeline job.
```

MarkEWaite commented on README.adoc:
```diff
@@ -51,6 +51,48 @@ The link:https://youtu.be/B_2FXWI6CWg[30 minute video clip] below introduces Mul
+
+The git plugin provides `Git Username and Password` binding that allows authenticated git operations over *HTTP* and *SSH* protocols using command line git in a Pipeline job.
+
+The git credential bindings are accessible through the `withCredentials` step of the link:https://plugins.jenkins.io/credentials-binding/[Credentials Binding] plugin.
```
Hyperlink to the step documentation?
```suggestion
The git credential bindings are accessible through the link:https://www.jenkins.io/doc/pipeline/steps/credentials-binding/#withcredentials-bind-credentials-to-variables[`withCredentials`] step of the link:https://plugins.jenkins.io/credentials-binding/[Credentials Binding] plugin.
```

MarkEWaite commented on README.adoc:
```diff
@@ -51,6 +51,48 @@ The link:https://youtu.be/B_2FXWI6CWg[30 minute video clip] below introduces Mul
+
+image:images/git-credentials-usernamepassword-binding-pipline-job.png[Git-Username-and Password-Binding-Pipeline-Job]
+
+NOTE: By default, two variable bindings, `GIT_USERNAME` and `GIT_PASSWORD`, bind with user’s credential i.e. username and password respectively, in a Pipeline jobs.
```
This seems like an interesting implementation detail, but placing it in a NOTE gives it more emphasis that I think is justified.  The user should not need to do anything with either of these variables.  They are internal implementation details.
```suggestion
Two variable bindings are used, `GIT_USERNAME` and `GIT_PASSWORD`, to pass the username and password to `sh`, `bat`, and `powershell` steps inside the `withCredentials` block of a Pipeline job.
```

MarkEWaite commented on README.adoc:
```diff
@@ -544,6 +586,25 @@ Project Name in ViewGit::
+
+To access the `Git Username and Password` binding in a pipeline job, visit <<credential-binding>>
+
+Incase of a FreeStyle project the steps are as follows:
```
Remove the 'Incase' word
```suggestion
Freestyle projects can use git credential binding with the following steps:
```

MarkEWaite commented on README.adoc:
```diff
@@ -544,6 +586,25 @@ Project Name in ViewGit::
+
+image:images/git-credentials-usernamepassword-binding-freestyle-project.png[Git-Username-and Password-Binding-Freestyle-project]
+
+NOTE: By default, two variable bindings, `GIT_USERNAME` and `GIT_PASSWORD`, bind with user’s credential i.e. username and password respectively, in a Pipeline jobs.
```
Remove the NOTE: emphasis here as well
```suggestion
Two variable bindings are used, `GIT_USERNAME` and `GIT_PASSWORD`, to pass the username and password to shell, batch, and powershell steps in a Freestyle job.
```

MarkEWaite reviewed (approved):
Minor phrasing changes proposed for your consideration.  Optional

MarkEWaite committed:
Minor documentation changes

rishabhBudhouliya commented on src/main/java/jenkins/plugins/git/GitCredentialBindings.java:
```diff
@@ -0,0 +1,86 @@
+            Node currentNode = buildExecutor.getOwner().getNode();
+            //Check node is not null
+            if (currentNode != null) {
+                GitTool nameSpecificGitTool = GitUtils.resolveGitTool(gitToolName,currentNode,new EnvVars(),listener);
```
nitpick: please follow conventional formatting rules (whitespace after comma)

rishabhBudhouliya commented on src/main/java/jenkins/plugins/git/GitUsernamePasswordBinding.java:
```diff
@@ -0,0 +1,208 @@
+            secretValues.put("GIT_ASKPASS", gitTempFile.getRemote());
+            return new MultiEnvironment(secretValues, publicValues, unbindTempDir.getUnbinder());
+        } else {
+            taskListener.getLogger().println("JGit and JGitApache type Git tools are not supported by this binding");
```
`getCliGitTool` could return a null value for a lot of other reasons than just the possibility of git tool being JGit and JGitApache types. I suggest you should make this log statement more general: 

"Unable to perform binding because git tool resolution has failed"

rishabhBudhouliya commented on src/main/java/jenkins/plugins/git/GitUsernamePasswordBinding.java:
```diff
@@ -0,0 +1,208 @@
+        StandardUsernamePasswordCredentials usernamePasswordCredentials = (StandardUsernamePasswordCredentials) credentials;
+        if(usernamePasswordCredentials.isUsernameSecret()){
+            secretValues.put(GIT_USERNAME_KEY, usernamePasswordCredentials.getUsername());
+        }else{
```
nitpick: formatting (whitespace)

rishabhBudhouliya commented on src/main/java/jenkins/plugins/git/GitUsernamePasswordBinding.java:
```diff
@@ -0,0 +1,208 @@
+    }
+
+    /*package*/void setGitEnvironmentVariables(@NonNull GitClient git, Map<String,String> publicValues) throws IOException, InterruptedException {
+        setGitEnvironmentVariables(git,null,publicValues);
```
nitpick: formatting

rishabhBudhouliya commented on src/main/java/jenkins/plugins/git/GitCredentialBindings.java:
```diff
@@ -0,0 +1,86 @@
+     * @param listener The task listener.
+     * @return a GitClient implementation {@link org.jenkinsci.plugins.gitclient.GitClient}
+     **/
+    GitClient getGitClientInstance(String gitExe, FilePath repository,
```
why is this added in the contract to be implemented? Can we not provide the logic assuming the implementation does not change for any reason?

rishabhBudhouliya replied:
similar to getCliGitTool

jglick replied:
(Still unexplained five years later.)

rishabhBudhouliya reviewed (approved):
I've left some fairly minor comments, nice work.
